--- rubygems-1.3.1/setup.rb~	2009-02-17 15:57:41.000000000 +0100
+++ rubygems-1.3.1/setup.rb	2009-02-17 15:57:51.077845298 +0100
@@ -148,8 +148,8 @@
   lib_dir_p = Pathname.new(lib_dir)
   bin_dir_p = Pathname.new(bin_dir)
 
-  lib_dir = File.join install_destdir, lib_dir_p.relative_path_from(top_dir)
-  bin_dir = File.join install_destdir, bin_dir_p.relative_path_from(top_dir)
+  lib_dir = File.join install_destdir, lib_dir_p
+  bin_dir = File.join install_destdir, bin_dir_p
 end
 
 mkdir_p lib_dir
